kge/util/sampler.py

```diff
@@ -188,6 +210,368 @@ def _filter_and_resample_fast(
+
+        # repeat scores as needed for WR sampling
+        if num_unique != self.num_samples:
+            scores = scores[
+                :, torch.cat((torch.arange(num_unique, device=scores.device), self._repeat_indexes))
+            ]
+        self.forward_time += time.time()
+
+        return scores
+
+    def to(self, device) -> "NaiveSharedNegativeSample":
+        super().to(device)
+        self._unique_samples = self._unique_samples.to(device)
+        self._repeat_indexes = self._repeat_indexes.to(device)
+        return self
+
+
+class DefaultSharedNegativeSample(BatchNegativeSample):
+    def __init__(
+        self,
+        config: Config,
+        configuration_key: str,
+        positive_triples: torch.Tensor,
+        slot: int,
+        num_samples: int,
+        unique_samples: torch.Tensor,
+        drop_index: torch.Tensor,
+        repeat_indexes: torch.Tensor,
+    ):
+        super().__init__(config, configuration_key, positive_triples, slot, num_samples)
+        self._unique_samples = unique_samples
+        self._drop_index = drop_index
+        self._repeat_indexes = repeat_indexes
+
+    def samples(self, indexes=None) -> torch.Tensor:
+        num_samples = self.num_samples
+        triples = (
+            self.positive_triples
+            if indexes is None
+            else self.positive_triples[indexes, :]
+        )
+        drop_index = self._drop_index if indexes is None else self._drop_index[indexes]
+        chunk_size = len(triples)
+
+        # create output tensor
+        device = self.positive_triples.device
+        num_unique = len(self._unique_samples) - 1
+        negative_samples = torch.empty(
+            chunk_size, num_unique, dtype=torch.long, device=device
+        )
+
+        # Add the first num_distinct samples for each positive. Dropping is
+        # performed by copying the last shared sample over the dropped sample
+        negative_samples[:, :] = self._unique_samples[:-1]
+        drop_rows = torch.nonzero(drop_index != num_unique, as_tuple=False).squeeze()
+        negative_samples[drop_rows, drop_index[drop_rows]] = self._unique_samples[-1]
+
+        # repeat indexes as needed for WR sampling
+        if num_unique != num_samples:
+            negative_samples = negative_samples[
+                :, torch.cat((torch.arange(num_unique), self._repeat_indexes))
+            ]
+
+        return negative_samples
+
+    def score(self, model, indexes=None) -> torch.Tensor:
+        if self._implementation != "batch":
+            return super().score(model, indexes)
+
+        # for batch, we have a faster implementation that avoids creating the full
+        # sample tensor
+        self.prepare_time = 0.0
+        self.forward_time = 0.0
+        slot = self.slot
+        unique_targets = self._unique_samples
+        num_unique = len(unique_targets) - 1
+        triples = (
+            self.positive_triples[indexes, :] if indexes else self.positive_triples
```

Comment:
We are loosing the time in the following few rows indexing the chunks.
Even if we are not chunking at all, we provide the score function with an index, which just indexes all triples.
Additionally we are doing this procedure num_slots x num_chunks times, which generates an overhead

Reply:
yes, we get roughly the same speed now